component/component.go

```diff
@@ -190,9 +190,54 @@ func (f CreateDefaultConfigFunc) CreateDefaultConfig() Config {
 	return f()
 }
 
-// InstanceID uniquely identifies a component instance
+// InstanceID uniquely identifies a component instance.
 type InstanceID struct {
-	ID          ID
-	Kind        Kind
-	PipelineIDs map[ID]struct{}
+	componentID ID
+	kind        Kind
+	pipelineIDs map[ID]struct{}
+}
+
+// NewInstanceID returns an ID that uniquely identifies a component.
+func NewInstanceID(componentID ID, kind Kind, pipelineIDs ...ID) *InstanceID {
+	instanceID := InstanceID{
+		componentID: componentID,
+		kind:        kind,
+		pipelineIDs: make(map[ID]struct{}, len(pipelineIDs)),
+	}
+	for _, pid := range pipelineIDs {
+		instanceID.pipelineIDs[pid] = struct{}{}
+	}
+	return &instanceID
+}
+
+// ComponentID returns the ComponentID associated with this instance.
+func (id *InstanceID) ComponentID() ID {
+	return id.componentID
+}
+
+// Kind returns the component Kind associated with this instance.
+func (id *InstanceID) Kind() Kind {
+	return id.kind
+}
+
+// PipelineIDs returns a set of PipelineIDs associated with this instance.
+func (id *InstanceID) PipelineIDs() map[ID]struct{} {
+	return id.pipelineIDs
+}
```

Comment:
Maybe it's over the top, but did you consider returning a copy of the map, to ensure the ID is truly immutable? Otherwise a caller can still modify the object.

I also wonder if it would be sensible to store the pipeline IDs in a way that makes the InstanceID type comparable. Again it doesn't seem to matter at the moment, since InstanceID is passed around by reference everywhere. Just thinking it could be good for future proofing.

Reply:
I had considered returning a copy of the map, but I wasn't sure it was worth the performance hit. These will be read often. This is an easy change to make though if anyone has strong opinions. Do you have a suggestion for how we might store PipelineIDs in a way that will make InstanceID comparable? If so, I'd be open to considering it. Bonus if it solves the true immutability problem as well.

Reply:
> I had considered returning a copy of the map, but I wasn't sure it was worth the performance hit. These will be read often. This is an easy change to make though if anyone has strong opinions.

Fair enough. I don't have a very strong opinion on this, just stumbled on the PR and thought it would be worth checking.

> Do you have a suggestion for how we might store PipelineIDs in a way that will make InstanceID comparable? If so, I'd be open to considering it. Bonus if it solves the true immutability problem as well.

One option would be to:
 - marshal the IDs, with some separator (space?) that's not permitted in type or name, and store as a string
 - change `PipelineIDs` to an iterator-type function like `pcommon.Map.Range` which incrementally unmarshals the IDs
 
The range function could be written in a way that works with Go 1.23's upcoming iterator functions: https://pkg.go.dev/iter#hdr-Iterators

Reply:
Both of these properties seem to me like things that could be added in the future in a backwards compatible way. If anything, we can document now that this is not comparable and that the map should not be modified, and leave adding this extra guarantees for a future PR.

Reply:
Seems fine to me. Having `map[ID]struct{}` in the exposed API doesn't seem ideal, and making it an iterator would be breaking again, but I don't think it's worth blocking over.

Reply:
> making it an iterator would be breaking again, but I don't think it's worth blocking over.

I don't feel like it's incompatible to add a new method for the iterator if we ensure the data is presented in the same way. I'll wait for Matt to give his opinion as well :)

Reply:
@axw, I refactored the InstanceID to be comparable per your suggestions. Take a look and see if this is what you had in mind. Let me know if it needs any further work. I chose the second (or third) best name for the iterator to leave the door open to rename it (to `AllPipelineIDs` or just `PipelineIDs`) when we can use the iterator functions coming in Go 1.23.